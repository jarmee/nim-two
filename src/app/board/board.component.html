<button
  type="button"
  data-test-id="execute-play-one"
  (click)="onExecutePlay(1)"
>
  1
</button>
<button
  type="button"
  data-test-id="execute-play-two"
  (click)="onExecutePlay(2)"
>
  2
</button>
<button
  type="button"
  data-test-id="execute-play-three"
  (click)="onExecutePlay(3)"
>
  3
</button>
<span>{{ amount$ | async }}</span>
<app-board-form></app-board-form>
