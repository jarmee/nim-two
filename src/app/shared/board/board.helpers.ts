import {
  Board,
  BoardDifference,
  BoardDifferences,
  Column,
  Columns,
  Path
} from "./board.model";

function createBoardDifferencesOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce(
      (flatten: [Column, Column][], element: any) => [...flatten, ...element],
      []
    );
}

function columnsWithEqualValues(difference: BoardDifference): boolean {
  return (
    !difference.newColumn ||
    !difference.currentColumn ||
    difference.newColumn.value !== difference.currentColumn.value
  );
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) return null;
  if (!path || !path.length) return null;

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) return null;
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  if (!boardA && !andBoardB) return [];

  return createBoardDifferencesOf(boardA, andBoardB).filter(
    columnsWithEqualValues
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(
  board: Board,
  withFilter: (column: Column) => boolean
): number {
  if (!board) return 0;

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce(
      (amount: number, columnCount: number) => (amount += columnCount),
      0
    );
}
