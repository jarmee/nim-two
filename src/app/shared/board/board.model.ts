export type Column = {
  value: boolean;
  player?: string;
};
export type Columns = Record<number, boolean>;
export type Board = {
  [key: number]: Columns;
  errorMessage: string;
};
