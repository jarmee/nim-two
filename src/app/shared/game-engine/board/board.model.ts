export enum BoardStatus {
  InProgress = 'IN_PROGRESS'
}
export type Column = {
  value: boolean;
  player?: string;
  errorMessage?: string;
};
export type Columns = Record<number, Column>;
export type Board = {
  [key: number]: Columns;
};
export type Path = string[];
export type BoardDifference = {
  newColumn: Column;
  currentColumn: Column;
  path: Path;
};
export type BoardDifferences = Array<BoardDifference>;
