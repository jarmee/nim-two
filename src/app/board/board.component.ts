import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { GameEngineFacade } from '../shared/game-engine';
import { Board } from '../shared/game-engine/board/board.model';
import { GameStatus } from '../shared/game-engine/state/state.model';
import { Player } from '../shared/game-engine/turn/turn.model';

@Component({
  selector: 'app-board',
  templateUrl: './board.component.html',
  styleUrls: ['./board.component.scss']
})
export class BoardComponent {
  amount$: Observable<number> = this.gameEngine.amount$;

  board$: Observable<Board> = this.gameEngine.board$;

  status$: Observable<GameStatus> = this.gameEngine.status$;

  selectedPlayer$: Observable<Player> = this.gameEngine.selectedPlayer$;

  messages$: Observable<string> = this.gameEngine.messages$;

  constructor(private gameEngine: GameEngineFacade) {}

  onExecutePlay(board: Board) {
    this.gameEngine.executePlay(board);
  }

  onReset() {
    this.gameEngine.reset();
  }
}
