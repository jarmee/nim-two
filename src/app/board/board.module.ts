import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { BoardComponent } from "./board.component";
import { BoardFormComponent } from "./form/board-form.component";

@NgModule({
  declarations: [BoardComponent, BoardFormComponent],
  imports: [CommonModule],
  exports: [BoardComponent]
})
export class BoardModule {}
