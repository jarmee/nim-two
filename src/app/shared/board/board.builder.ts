import { Board } from "./board.model";

export class BoardBuilder {
  board: Board = {};

  addRowWithColumns(...columns: boolean[]): BoardBuilder {
    const currentIndex = Object.keys(this.board).length++;
    this.board = {
      ...this.board,
      [currentIndex]: {
        ...columns
          .map((value: boolean, index: number) => ({
            [index]: {
              value,
              player: null
            }
          }))
          .reduce((row: any, column: any) => ({ ...row, ...column }), {})
      }
    };
    return this;
  }

  build(): Board {
    return this.board;
  }

  static create(): BoardBuilder {
    return new BoardBuilder();
  }
}
