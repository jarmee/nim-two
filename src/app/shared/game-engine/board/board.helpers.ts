import { Board, BoardDifference, BoardDifferences, Column, Columns, Path } from './board.model';

function createBoardDifferencesOf(boardA: Board, andBoardB: Board): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce((flatten: [Column, Column][], element: any) => [...flatten, ...element], []);
}

export function areColumnsDifferentByValue(newColumn: Column, currentColumn: Column): boolean {
  if (!newColumn && !currentColumn) {
    return false;
  }
  if (!newColumn || !currentColumn) {
    return true;
  }
  return newColumn.value !== currentColumn.value;
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) {
    return null;
  }
  if (!path || !path.length) {
    return null;
  }

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) {
      return null;
    }
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(boardA: Board, andBoardB: Board): BoardDifferences {
  if (!boardA && !andBoardB) {
    return [];
  }

  return createBoardDifferencesOf(boardA, andBoardB).filter(({ newColumn, currentColumn }: BoardDifference) =>
    areColumnsDifferentByValue(newColumn, currentColumn)
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(board: Board, withFilter: (column: Column) => boolean): number {
  if (!board) {
    return 0;
  }

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce((amount: number, columnCount: number) => (amount += columnCount), 0);
}

export function exchangeColumnsOf(
  board: Board,
  exchangeColumnCallback: (column: Column, path?: Path) => Column
): Board {
  if (!board) {
    return null;
  }

  return Object.keys(board)
    .map(rowKey => ({
      [rowKey]: Object.keys(board[rowKey])
        .map(columnKey => ({
          [columnKey]: exchangeColumnCallback
            ? exchangeColumnCallback(board[rowKey][columnKey], [rowKey, columnKey])
            : { ...board[rowKey][columnKey] }
        }))
        .reduce((newRow, column) => ({ ...newRow, ...column }))
    }))
    .reduce((newBoard, row) => ({ ...newBoard, ...row }), {});
}
