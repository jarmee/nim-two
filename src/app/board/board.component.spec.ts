import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { BoardComponent } from "./board.component";
import { By } from "@angular/platform-browser";

describe("BoardComponent", () => {
  let component: BoardComponent;
  let fixture: ComponentFixture<BoardComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [BoardComponent]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(BoardComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  describe("onExecutePlay", () => {
    describe("", () => {
      it("should call the onExecutePlay event handler on click", () => {
        const spyOnOnExecutePlay = jest.spyOn(component, "onExecutePlay");
        const buttonElement = fixture.debugElement.query(
          By.css("[data-test-id='execute-play-one']")
        );

        buttonElement.triggerEventHandler("click", {});

        expect(spyOnOnExecutePlay).toHaveBeenCalledWith(1);
      });
    });
  });
});
