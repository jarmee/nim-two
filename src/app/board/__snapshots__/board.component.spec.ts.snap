// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardComponent should match the snapshot 1`] = `
<app-board
  amount$={[Function AnonymousSubject]}
  board$={[Function AnonymousSubject]}
  gameEngine={[Function GameEngineFacade]}
  messages$={[Function AnonymousSubject]}
  selectedPlayer$={[Function AnonymousSubject]}
  status$={[Function AnonymousSubject]}
>
  <div
    class="container"
  >
    <div
      class="row mt-5"
    >
      <div
        class="col-12 text-center"
      >
        <h1
          class="col-12"
        >
          You vs ðŸ¤–
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      ðŸ¤– 
    </div>
    <div
      class="row mt-5"
    >
      <div
        class="col-12"
      >
        <app-board-form
          data-test-id="board"
          ng-reflect-board="[object Object]"
        >
          <form
            class="ng-untouched ng-pristine ng-valid"
            ng-reflect-form="[object Object]"
            novalidate=""
          >
            
            
            
            <div
              class="d-flex justify-content-around ng-untouched ng-pristine ng-valid"
              ng-reflect-name="0"
            >
              
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="0"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-0"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-0"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-0"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="1"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-1"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-1"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-1"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="2"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-2"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-2"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-2"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="3"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-3"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-3"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-3"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="4"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-4"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-4"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-4"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="5"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-5"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-5"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-5"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="6"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-6"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-6"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-6"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="7"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-7"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-7"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-7"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="8"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-8"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-8"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-8"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="9"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-9"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-9"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-9"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="10"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-10"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-10"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-10"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="11"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-11"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-11"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-11"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
              <div
                class="ng-untouched ng-pristine ng-valid"
                ng-reflect-name="12"
              >
                <app-match-control
                  class="ng-untouched ng-pristine ng-valid"
                  formcontrolname="value"
                  ng-reflect-form-control-name="value"
                  ng-reflect-id="0-12"
                  ng-reflect-name="value"
                  ng-reflect-ng-class="[object Object]"
                >
                  <div
                    class="custom-control custom-checkbox"
                  >
                    <input
                      class="custom-control-input"
                      data-test-id="match-checkbox"
                      id="value-0-12"
                      name="value"
                      type="checkbox"
                    />
                    <label
                      class="custom-control-label"
                      for="value-0-12"
                    />
                  </div>
                </app-match-control>
                <app-player-batch
                  ng-reflect-name=""
                >
                  
                </app-player-batch>
              </div>
            </div>
            <div />
            
            <div
              class="row mt-5"
            >
              <button
                class="btn btn-primary"
                data-test-id="execute-play-button"
                type="button"
              >
                 Finish Play 
              </button>
              <button
                class="btn btn-danger"
                data-test-id="reset-button"
                type="button"
              >
                 Reset Game 
              </button>
            </div>
          </form>
        </app-board-form>
      </div>
    </div>
  </div>
</app-board>
`;
